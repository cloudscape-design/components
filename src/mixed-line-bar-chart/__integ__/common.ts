// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { BasePageObject } from '@cloudscape-design/browser-test-tools/page-objects';
import useBrowser from '@cloudscape-design/browser-test-tools/use-browser';

import createWrapper from '../../../lib/components/test-utils/selectors';

export class MixedChartPage extends BasePageObject {
  currentIndex: number | undefined;
  wrapper = createWrapper().findMixedLineBarChart();

  constructor(browser: ConstructorParameters<typeof BasePageObject>[0]) {
    super(browser);
  }

  getPopoverHeaderText() {
    return this.getText(this.wrapper.findDetailPopover().findHeader().toSelector());
  }

  async getKeyValuePairsText() {
    return (await this.getElementsText(this.wrapper.findDetailPopover().findSeries().toSelector())).join('\n');
  }

  async navigateToDatum(index: number, interaction: string) {
    const barGroup = this.wrapper.findBarGroups().get(index).toSelector();
    if (interaction === 'hover') {
      return this.hoverElement(barGroup);
    } else {
      // If a popover was pinned, we need to unpin it before pinning another one.
      if (this.currentIndex) {
        await this.click(this.wrapper.findDetailPopover().findDismissButton().toSelector());
      }
      this.currentIndex = index;
      await this.clickBarGroup(barGroup);
    }
  }

  async clickBarGroup(selector: string) {
    // We need to use `page.buttonDownOnElement` and `page.buttonUp` instead of `page.click` because the ancestor SVG element
    // intercepts the click and this is seen by Webdriver as an error and thrown as such, although it works for us
    // (the component manages the event accordingly as coming from the corresponding bar group).
    await this.buttonDownOnElement(selector);
    await this.buttonUp();
  }
}

export function setupTest(url: string, testFn: (page: MixedChartPage) => Promise<void>) {
  return useBrowser(async browser => {
    const page = new MixedChartPage(browser);
    await browser.url(url);
    await testFn(page);
  });
}
