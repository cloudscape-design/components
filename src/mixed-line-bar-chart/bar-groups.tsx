// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import React from 'react';

import { ChartDataTypes } from './interfaces';
import { ScaledBarGroup } from './make-scaled-bar-groups';

import styles from './styles.css.js';

const OUTLINE_PADDING = 2;

export interface BarGroups<T extends ChartDataTypes> {
  ariaLabel: string;
  isRefresh: boolean;
  isPopoverPinned: boolean;
  barGroups: ScaledBarGroup<T>[];
  highlightedGroupIndex: null | number;
  highlightedGroupRef: React.RefObject<SVGRectElement>;
}

/**
 * The component renders empty SVG rectangles corresponding to the bar group slots.
 * The highlighted group rectangle is used for the pseudo-focus and therefore requires ARIA attributes.
 * Other rectangles are only needed for the "findBarGroups" test-utils selector.
 */
export default function BarGroups<T extends ChartDataTypes>({
  ariaLabel,
  isRefresh,
  isPopoverPinned,
  barGroups,
  highlightedGroupIndex,
  highlightedGroupRef,
}: BarGroups<T>) {
  return (
    <g role="group" aria-hidden="true">
      {barGroups.map((group, index) => (
        <rect
          key={index}
          x={Math.max(0, group.position.x - OUTLINE_PADDING)}
          y={Math.max(0, group.position.y - OUTLINE_PADDING)}
          width={group.position.width + 2 * OUTLINE_PADDING}
          height={group.position.height + 2 * OUTLINE_PADDING}
          ref={index === highlightedGroupIndex ? highlightedGroupRef : undefined}
          rx={isRefresh ? 4 : 2}
          role="button"
          aria-label={ariaLabel}
          aria-haspopup={true}
          aria-expanded={isPopoverPinned}
          fill="none"
          className={styles['bar-group']}
        />
      ))}
    </g>
  );
}
