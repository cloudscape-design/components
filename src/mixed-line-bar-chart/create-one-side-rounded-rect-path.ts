/* eslint-disable header/header */
// Source https://stackoverflow.com/a/65186378/16648714

export function createOneSideRoundedRectPath(
  { x, y, width, height }: { x: number; y: number; width: number; height: number },
  radius: number,
  side: 'left' | 'right' | 'top' | 'bottom'
) {
  const coordinates = [
    { x, y },
    { x: x + width, y },
    { x: x + width, y: y + height },
    { x: x, y: y + height },
  ];

  // Starting from the target side.
  const startIndex = { left: 2, right: 0, top: 3, bottom: 1 }[side];

  let path = '';
  for (let i = startIndex; i < startIndex + coordinates.length + 1; i++) {
    const start = coordinates[i % coordinates.length];
    const end = coordinates[(i + 1) % coordinates.length];

    // Define curvature for the first 3 points to only add rounded corners to the first side.
    const c = radius && i < startIndex + 3 ? Math.min(radius / Math.hypot(end.x - start.x, end.y - start.y), 0.5) : 0;

    // Insert start point. When c=0 the start point is [start.x, start.y].
    if (i === startIndex) {
      path += `M${start.x * (1 - c) + end.x * c},${start.y * (1 - c) + end.y * c}`;
    }
    // Insert quadratic curve.
    if (i > startIndex) {
      path += `Q${start.x},${start.y} ${start.x * (1 - c) + end.x * c},${start.y * (1 - c) + end.y * c}`;
    }
    // Draw line to the end point. When c=0 the end point is [end.x, end.y].
    if (i < startIndex + coordinates.length) {
      path += `L${start.x * c + end.x * (1 - c)},${start.y * c + end.y * (1 - c)}`;
    }
  }
  return path + 'Z';
}
