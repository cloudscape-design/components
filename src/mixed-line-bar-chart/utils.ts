// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { ChartSeriesMarkerType } from '../internal/components/chart-series-marker';
import { ChartDataTypes, InternalChartSeries, MixedLineBarChartProps } from './interfaces';
import { ScaledBarGroup } from './make-scaled-bar-groups';

export const chartLegendMap: Record<string, ChartSeriesMarkerType> = {
  line: 'line',
  bar: 'rectangle',
  threshold: 'dashed',
};

// Starting from the given index, find the first x value in the x domain that has bar data attached to it.
export const nextValidDomainIndex = <T>(nextGroupIndex: number, barGroups: ScaledBarGroup<T>[], direction = 1) => {
  let index = nextGroupIndex;

  if (index < 0 || index >= barGroups.length) {
    index = 0;
  }

  do {
    if (barGroups[index].isValid && barGroups[index].hasData) {
      return index;
    }
    index += direction;

    // Loop back to the beginning if necessary
    if (index >= barGroups.length) {
      index = 0;
    } else if (index < 0) {
      index = barGroups.length - 1;
    }
  } while (index !== nextGroupIndex);
  return 0;
};

/**
 * Find the subset of series that are individually navigable with keyboard.
 * Lines and thresholds are navigated individually, while bar series are grouped as one.
 */
export function findNavigableSeries<T extends ChartDataTypes>(series: ReadonlyArray<InternalChartSeries<T>>) {
  const navigableSeries: Array<MixedLineBarChartProps.ChartSeries<T>> = [];
  let navigableBarSeriesIndex = -1;

  series.forEach(internalSeries => {
    if (internalSeries.series.type === 'bar') {
      // Only include the first bar series because all bar series are handled as one
      if (navigableBarSeriesIndex === -1) {
        navigableBarSeriesIndex = navigableSeries.length;
        navigableSeries.push(internalSeries.series);
      }
    } else {
      navigableSeries.push(internalSeries.series);
    }
  });
  return { navigableSeries, navigableBarSeriesIndex };
}

/**
 * Checks if two x values are equal.
 * With a special treat for Date values which need to be converted to numbers first.
 */
export const matchesX = <T>(x1: T, x2: T) => {
  if (x1 instanceof Date && x2 instanceof Date) {
    return x1.getTime() === x2.getTime();
  }
  return x1 === x2;
};

export type StackedBarValues = Map<string | number, Map<number, number>>;

// Unlike for regular bars, stacked bar series values depend on the predecessors.
// The function computes all stacked values grouped by X and series index.
export function calculateStackedBarValues(
  dataBySeries: Array<readonly MixedLineBarChartProps.Datum<ChartDataTypes>[]>
): StackedBarValues {
  const negativeValues = new Map<string | number, number>();
  const positiveValues = new Map<string | number, number>();
  const values = new Map<string | number, Map<number, number>>();
  for (let seriesIndex = 0; seriesIndex < dataBySeries.length; seriesIndex++) {
    for (const datum of dataBySeries[seriesIndex]) {
      const key = getKeyValue(datum.x);
      if (datum.y < 0) {
        negativeValues.set(key, (negativeValues.get(key) ?? 0) + datum.y);
      } else {
        positiveValues.set(key, (positiveValues.get(key) ?? 0) + datum.y);
      }
      const seriesValue = (datum.y < 0 ? negativeValues.get(key) : positiveValues.get(key)) ?? 0;
      const valuesByIndex = values.get(key) ?? new Map<number, number>();
      valuesByIndex.set(seriesIndex, seriesValue);
      values.set(key, valuesByIndex);
    }
  }
  return values;
}

/** Returns string or number value for ChartDataTypes key */
export const getKeyValue = (key: ChartDataTypes) => (key instanceof Date ? key.getTime() : key);

export function isYThreshold<T>(
  series: MixedLineBarChartProps.ChartSeries<T>
): series is MixedLineBarChartProps.YThresholdSeries {
  return series.type === 'threshold' && 'y' in series;
}

export function isXThreshold<T>(
  series: MixedLineBarChartProps.ChartSeries<T>
): series is MixedLineBarChartProps.XThresholdSeries<T> {
  return series.type === 'threshold' && 'x' in series;
}

export function isDataSeries<T>(
  series: MixedLineBarChartProps.ChartSeries<T>
): series is MixedLineBarChartProps.DataSeries<T> {
  return series.type === 'line' || series.type === 'bar';
}

// Source https://stackoverflow.com/a/65186378/16648714
export function createOneSideRoundedRectPath(
  { x, y, width, height }: { x: number; y: number; width: number; height: number },
  radius: number,
  side: 'left' | 'right' | 'top' | 'bottom'
) {
  const coordinates = [
    { x, y },
    { x: x + width, y },
    { x: x + width, y: y + height },
    { x: x, y: y + height },
  ];

  // Starting from the target side.
  const startIndex = { left: 2, right: 0, top: 3, bottom: 1 }[side];

  let path = '';
  for (let i = startIndex; i < startIndex + coordinates.length + 1; i++) {
    const start = coordinates[i % coordinates.length];
    const end = coordinates[(i + 1) % coordinates.length];

    // Define curvature for the first 3 points to only add rounded corners to the first side.
    const c = radius && i < startIndex + 3 ? Math.min(radius / Math.hypot(end.x - start.x, end.y - start.y), 0.5) : 0;

    // Insert start point. When c=0 the start point is [start.x, start.y].
    if (i === startIndex) {
      path += `M${start.x * (1 - c) + end.x * c},${start.y * (1 - c) + end.y * c}`;
    }
    // Insert quadratic curve.
    if (i > startIndex) {
      path += `Q${start.x},${start.y} ${start.x * (1 - c) + end.x * c},${start.y * (1 - c) + end.y * c}`;
    }
    // Draw line to the end point. When c=0 the end point is [end.x, end.y].
    if (i < startIndex + coordinates.length) {
      path += `L${start.x * c + end.x * (1 - c)},${start.y * c + end.y * (1 - c)}`;
    }
  }
  return path + 'Z';
}
