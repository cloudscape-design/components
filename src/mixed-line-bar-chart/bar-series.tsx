// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import React, { useMemo } from 'react';
import clsx from 'clsx';
import { ScaleContinuousNumeric, ScaleTime } from 'd3-scale';

import { ChartScale, NumericChartScale } from '../internal/components/cartesian-chart/scales';
import { ChartDataTypes, MixedLineBarChartProps } from './interfaces';
import { matchesX, getKeyValue, StackedOffsets } from './utils';
import styles from './styles.css.js';
import { useVisualRefresh } from '../internal/hooks/use-visual-mode';
export interface BarSeriesProps<T> {
  axis: 'x' | 'y';

  series: MixedLineBarChartProps.DataSeries<T>;

  color: string;
  totalSeriesCount: number;
  seriesIndex: number;

  xScale: ChartScale;
  yScale: NumericChartScale;

  plotSize: number;
  chartAreaClipPath: string;

  highlighted: boolean;
  dimmed: boolean;
  highlightedGroupIndex: number | null;

  // Contains the cumulative offset for each x value in a stacked bar chart
  stackedBarOffsets?: StackedOffsets;
}

export default function BarSeries<T extends ChartDataTypes>({
  axis,
  series,
  color,
  xScale,
  yScale,
  highlighted,
  dimmed,
  highlightedGroupIndex,
  stackedBarOffsets,
  totalSeriesCount,
  seriesIndex,
  plotSize,
  chartAreaClipPath,
}: BarSeriesProps<T>) {
  const isRefresh = useVisualRefresh();

  const xCoordinates = useMemo(() => {
    if (series.type !== 'bar' || !xScale.isCategorical()) {
      return [];
    }

    const yContinuosScale: ScaleContinuousNumeric<number, number> | ScaleTime<number, number> = yScale.d3Scale;
    const xPoints = (series.data as ReadonlyArray<MixedLineBarChartProps.Datum<ChartDataTypes>>).map(
      ({ x }) => xScale.d3Scale(x) || NaN
    );

    const zeroY = yScale.d3Scale(0) ?? NaN;
    const baseY = isFinite(zeroY) ? Math.min(plotSize, zeroY) : plotSize;

    let barWidth = xScale.d3Scale.bandwidth();

    const PADDING = 4;
    const MINWIDTH = 4;

    if (!stackedBarOffsets && totalSeriesCount > 1) {
      // Regular grouped bars
      barWidth = (barWidth - (totalSeriesCount - 1) * PADDING) / totalSeriesCount;
      barWidth = Math.max(barWidth, MINWIDTH);
    }

    return xPoints.map((x, i) => {
      const d = series.data[i];
      let barX = x;
      let yValue = d.y;

      if (stackedBarOffsets) {
        // Stacked bars
        const offsetMap = d.y < 0 ? stackedBarOffsets.negativeOffsets : stackedBarOffsets.positiveOffsets;
        yValue = d.y + (offsetMap[getKeyValue(d.x)] || 0);
      } else if (totalSeriesCount > 1) {
        // Regular grouped bars
        barX += seriesIndex * (barWidth + PADDING);
      }

      // Account for negative values growing "down" instead of "up"
      yValue = yValue < 0 ? yValue - d.y : yValue;

      return {
        x: barX,
        y: yContinuosScale(yValue) ?? NaN,
        width: barWidth,
        height: Math.abs((yContinuosScale(d.y) ?? NaN) - baseY),
      };
    });
  }, [series, xScale, yScale, plotSize, stackedBarOffsets, totalSeriesCount, seriesIndex]);

  const highlightedXValue = highlightedGroupIndex !== null ? xScale.domain[highlightedGroupIndex] : null;

  return (
    <g
      aria-label={series.title}
      clipPath={`url(#${chartAreaClipPath})`}
      className={clsx(styles.series, styles['series--bar'], {
        [styles['series--highlighted']]: highlighted,
        [styles['series--dimmed']]: dimmed,
      })}
    >
      {xCoordinates.map(
        ({ x, y, width, height }, i) =>
          isFinite(x) &&
          isFinite(height) && (
            <rect
              key={`bar-${i}`}
              fill={color}
              x={axis === 'x' ? x : y - height}
              y={axis === 'x' ? y : x}
              width={axis === 'x' ? width : height}
              height={axis === 'x' ? height : width}
              rx={isRefresh ? '4px' : '0px'}
              className={clsx(styles.series__rect, {
                [styles['series--dimmed']]:
                  highlightedXValue !== null && !matchesX(highlightedXValue, series.data[i].x),
              })}
            />
          )
      )}
    </g>
  );
}
