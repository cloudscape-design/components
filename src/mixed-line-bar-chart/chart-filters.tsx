// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import React, { useMemo } from 'react';
import clsx from 'clsx';

import InternalSpaceBetween from '../space-between/internal';
import Filter, { ChartFilterProps } from '../internal/components/chart-filter';

import { ChartDataTypes, MixedLineBarChartProps, InternalChartSeries } from './interfaces';
import { chartLegendMap } from './utils';
import styles from './styles.css.js';

interface InternalChartFiltersProps<T extends ChartDataTypes> {
  series: ReadonlyArray<InternalChartSeries<T>>;
  visibleSeries: ReadonlyArray<MixedLineBarChartProps.ChartSeries<T>>;
  onChange: (selectedSeries: ReadonlyArray<MixedLineBarChartProps.ChartSeries<T>>) => void;
  i18nStrings: MixedLineBarChartProps<T>['i18nStrings'];
  hideFilter?: boolean;
  additionalFilters?: React.ReactNode;
}

export default function InternalChartFilters<T extends number | string | Date>({
  series,
  visibleSeries,
  onChange,
  i18nStrings,
  hideFilter,
  additionalFilters,
}: InternalChartFiltersProps<T>) {
  const filterItems: ChartFilterProps<MixedLineBarChartProps.ChartSeries<T>>['series'] = useMemo(
    () =>
      series.map(({ series, color }) => ({
        label: series.title,
        type: chartLegendMap[series.type],
        color,
        datum: series,
      })),
    [series]
  );

  return (
    <InternalSpaceBetween
      size="l"
      direction="horizontal"
      className={clsx({
        [styles['has-default-filter']]: !hideFilter,
      })}
    >
      {!hideFilter && (
        <Filter series={filterItems} onChange={onChange} selectedSeries={visibleSeries} i18nStrings={i18nStrings} />
      )}
      {additionalFilters}
    </InternalSpaceBetween>
  );
}
