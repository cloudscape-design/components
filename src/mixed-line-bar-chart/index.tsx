// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import React from 'react';

import useBaseComponent from '../internal/hooks/use-base-component';
import { applyDisplayName } from '../internal/utils/apply-display-name';
import { MixedLineBarChartProps } from './interfaces';
import InternalMixedLineBarChart from './internal';

export { MixedLineBarChartProps };

function MixedLineBarChart<T extends number | string | Date>({
  series = [],
  height = 500,
  xScaleType = 'linear',
  yScaleType = 'linear',
  stackedBars = false,
  horizontalBars = false,
  statusType = 'finished',
  detailPopoverSize = 'medium',
  emphasizeBaselineAxis = true,
  ...props
}: MixedLineBarChartProps<T>) {
  const baseComponentProps = useBaseComponent('MixedLineBarChart', {
    props: {
      detailPopoverSize,
      emphasizeBaselineAxis,
      fitHeight: props.fitHeight,
      hideFilter: props.hideFilter,
      hideLegend: props.hideLegend,
      horizontalBars,
      stackedBars,
      xScaleType,
      yScaleType,
    },
  });
  return (
    <InternalMixedLineBarChart
      series={series}
      height={height}
      xScaleType={xScaleType}
      yScaleType={yScaleType}
      stackedBars={stackedBars}
      horizontalBars={horizontalBars}
      statusType={statusType}
      detailPopoverSize={detailPopoverSize}
      emphasizeBaselineAxis={emphasizeBaselineAxis}
      {...props}
      {...baseComponentProps}
    />
  );
}

applyDisplayName(MixedLineBarChart, 'MixedLineBarChart');

export default MixedLineBarChart;
