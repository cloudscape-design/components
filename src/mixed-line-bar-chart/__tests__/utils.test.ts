// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { matchesX, calculateStackedBarValues } from '../../../lib/components/mixed-line-bar-chart/utils';
import { createOneSideRoundedRectPath } from '../../../lib/components/mixed-line-bar-chart/create-one-side-rounded-rect-path';

import { barSeries, barSeries2 } from './common';

describe('matchesX', () => {
  test('with numbers', () => {
    expect(matchesX(5, 10)).toBeFalsy();
    expect(matchesX(10, 5)).toBeFalsy();

    expect(matchesX(33, 33)).toBeTruthy();

    expect(matchesX(20, null)).toBeFalsy();
  });

  test('with strings', () => {
    expect(matchesX('test', 'foo')).toBeFalsy();
    expect(matchesX('foo', 'test')).toBeFalsy();

    expect(matchesX('bar', 'bar')).toBeTruthy();

    expect(matchesX('foo', null)).toBeFalsy();
  });

  test('with dates', () => {
    const date1 = new Date(2020, 10, 5);
    const date2 = new Date(2020, 10, 7);

    expect(matchesX(date1, date2)).toBeFalsy();
    expect(matchesX(date2, date1)).toBeFalsy();

    expect(matchesX(date1, date1)).toBeTruthy();
    expect(matchesX(date1, new Date(+date1))).toBeTruthy();

    expect(matchesX(date1, null)).toBeFalsy();
  });
});

describe('calculateStackedBarValues', () => {
  test('without data', () => {
    expect(calculateStackedBarValues([])).toEqual(new Map());
  });

  test('with categorical data', () => {
    const data = [barSeries.data, barSeries2.data, barSeries.data];
    const computed = calculateStackedBarValues(data);
    expect(computed).toEqual(
      new Map([
        [
          'Potatoes',
          new Map([
            [0, 77],
            [1, 77 + 10],
            [2, 77 + 10 + 77],
          ]),
        ],
        [
          'Chocolate',
          new Map([
            [0, 546],
            [1, 546 + 20],
            [2, 546 + 20 + 546],
          ]),
        ],
        [
          'Apples',
          new Map([
            [0, 52],
            [2, 52 + 0 + 52],
          ]),
        ],
        [
          'Oranges',
          new Map([
            [0, 47],
            [1, 47 + 50],
            [2, 47 + 50 + 47],
          ]),
        ],
      ])
    );
  });

  test('with timeseries data', () => {
    const date = new Date('1995-12-17T03:24:00');
    const data = [[{ x: date, y: 1 }], [{ x: date, y: 2 }], [{ x: date, y: 3 }]];
    const computed = calculateStackedBarValues(data);
    expect(computed).toEqual(
      new Map([
        [
          date.getTime(),
          new Map([
            [0, 1],
            [1, 1 + 2],
            [2, 1 + 2 + 3],
          ]),
        ],
      ])
    );
  });

  test('with numeric data', () => {
    const data = [[{ x: 1, y: 1 }], [{ x: 1, y: 2 }], [{ x: 1, y: 3 }]];
    const computed = calculateStackedBarValues(data);
    expect(computed).toEqual(
      new Map([
        [
          1,
          new Map([
            [0, 1],
            [1, 1 + 2],
            [2, 1 + 2 + 3],
          ]),
        ],
      ])
    );
  });

  test('with mixed positive and negative numbers', () => {
    const data = [
      [
        { x: 1, y: -1 },
        { x: 2, y: 3 },
        { x: 3, y: -4 },
      ],
      [
        { x: 1, y: 2 },
        { x: 2, y: -3 },
        { x: 3, y: -4 },
      ],
      [
        { x: 1, y: 2 },
        { x: 2, y: -1 },
        { x: 3, y: 10 },
      ],
    ];
    const computed = calculateStackedBarValues(data);
    expect(computed).toEqual(
      new Map([
        [
          1,
          new Map([
            [0, -1],
            [1, 2],
            [2, 2 + 2],
          ]),
        ],
        [
          2,
          new Map([
            [0, 3],
            [1, -3],
            [2, -3 + -1],
          ]),
        ],
        [
          3,
          new Map([
            [0, -4],
            [1, -4 + -4],
            [2, 10],
          ]),
        ],
      ])
    );
  });
});

describe('createOneSideRoundedRectPath', () => {
  test.each(['left', 'right', 'top', 'bottom'] as const)('creates path with rx=0 and side="%s"', side => {
    const path = createOneSideRoundedRectPath({ x: 1, y: 2, width: 3, height: 4 }, 0, side);
    expect(path).toMatchSnapshot();
  });

  test.each(['left', 'right', 'top', 'bottom'] as const)('creates path with rx=4 and side="%s"', side => {
    const path = createOneSideRoundedRectPath({ x: 1, y: 2, width: 3, height: 4 }, 0, side);
    expect(path).toMatchSnapshot();
  });
});
