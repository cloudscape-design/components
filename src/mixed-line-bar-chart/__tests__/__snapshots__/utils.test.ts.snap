// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createOneSideRoundedRectPath creates path with rx=0 and side="bottom" 1`] = `"M4,2L4,6Q4,6 4,6L1,6Q1,6 1,6L1,2Q1,2 1,2L4,2Q4,2 4,2Z"`;

exports[`createOneSideRoundedRectPath creates path with rx=0 and side="left" 1`] = `"M4,6L1,6Q1,6 1,6L1,2Q1,2 1,2L4,2Q4,2 4,2L4,6Q4,6 4,6Z"`;

exports[`createOneSideRoundedRectPath creates path with rx=0 and side="right" 1`] = `"M1,2L4,2Q4,2 4,2L4,6Q4,6 4,6L1,6Q1,6 1,6L1,2Q1,2 1,2Z"`;

exports[`createOneSideRoundedRectPath creates path with rx=0 and side="top" 1`] = `"M1,6L1,2Q1,2 1,2L4,2Q4,2 4,2L4,6Q4,6 4,6L1,6Q1,6 1,6Z"`;

exports[`createOneSideRoundedRectPath creates path with rx=4 and side="bottom" 1`] = `"M4,2L4,6Q4,6 4,6L1,6Q1,6 1,6L1,2Q1,2 1,2L4,2Q4,2 4,2Z"`;

exports[`createOneSideRoundedRectPath creates path with rx=4 and side="left" 1`] = `"M4,6L1,6Q1,6 1,6L1,2Q1,2 1,2L4,2Q4,2 4,2L4,6Q4,6 4,6Z"`;

exports[`createOneSideRoundedRectPath creates path with rx=4 and side="right" 1`] = `"M1,2L4,2Q4,2 4,2L4,6Q4,6 4,6L1,6Q1,6 1,6L1,2Q1,2 1,2Z"`;

exports[`createOneSideRoundedRectPath creates path with rx=4 and side="top" 1`] = `"M1,6L1,2Q1,2 1,2L4,2Q4,2 4,2L4,6Q4,6 4,6L1,6Q1,6 1,6Z"`;
